### PR TITLE
Link: design changes (part II externalLinkIcon) 

### PR DESCRIPTION
**Pull request by AlbertCarreras**
#### What changed?

Link: design changes (part II: externalLinkIcon)

**Main changes** 
 new `externalLinkIcon` prop with Flow type `"none" \| "default" \| {\| size: $ElementType<React$ElementConfig<typeof Text>, "size">, color: $ElementType<React$ElementConfig<typeof Icon>, "color">, \|} |  
Screen Shot 2022-04-28 at 3 44 25 PM

#### Why?

Follow up PR for https://github.com/pinterest/gestalt/pull/2073
This is the step 2 of the [Link standardization project](https://paper.dropbox.com/doc/Gestalt-Link-Standardization--BgHgy3V31Efv1FE0GiNiKd6RAg-Opk3rapNVzU43gEfjedvb).

### Links

- [Jira](https://jira.pinadmin.com/browse/GESTALT-XXXX)
- [TDD](https://paper.dropbox.com/doc/Link--BgHiqMVz~DGCTqGCYy0GzE~_Ag-wVPcHLCkPfUVn8mIDtIN9)
- [Figma](https://www.figma.com/file/oUxggN7E1NlZftV4G4UfJD/Links-exploration?node-id=132%3A95

### Checklist

- [x] Added unit and Flow Tests
- [x] Added documentation + accessibility tests
- [x] Verified accessibility: keyboard & screen reader interaction
- [x] Checked dark mode, responsiveness, and right-to-left support
- [x] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 ### Major
 
-- Link: design changes (part I: inline/underline) (#2073)
+- Link: design changes (part I inline/underline) (#2073)
```
Comment (dangerismycat) [inline CHANGELOG.md:11]: This is typically a generated file — is this change intentional?

File: docs/pages/link.js [hunk 2]
```diff
@@ -233,14 +233,16 @@ export default function DocsPage({ generatedDocGen }: {| generatedDocGen: DocGen
...
         </MainSection.Subsection>
-        {/* <MainSection.Subsection columns={2}>
+        <MainSection.Subsection columns={2}>
           <MainSection.Card
             type="do"
-            description={`\`isExternal\` PROP COMING SOON.
-
-Use the "visit" icon to represent an external Link/domain in a text context. The icon should match the text size. Please note: Disclaimers and links taking users to Pinterest sub-sites don't need to display an external link icon.`}
+            description={`
+Use the "visit" icon to represent an external Link/domain in a text context. The icon should match the text size and color. Please note: Disclaimers and links taking users to Pinterest sub-sites don't need to display an external link icon.`}
```
Comment (dangerismycat) [inline docs/pages/link.js:240]: I'm not sure I understand the word "disclaimers" in this context?

File: docs/pages/link.js [hunk 3]
```diff
@@ -430,8 +432,20 @@ Accessible content is critical if we consider that assistive technology also pre
...
         <MainSection.Subsection
-          title="Accessible Tab Link"
+          title="External links"
+          description={`An external link, also called an outbound link, is a link from Pinterest to a different website.
+
+When rendering an external Link, add text for screen readers to announce that Link will go to a different destination. Users should be informed that they will be moving out of a domain and which domain they are moving to. This is particularly relevant for those with cognitive impairments or people relying on assistive technology.
+
+When possible, limit one external Link per paragraph, as adding more than two icons in the same block of text can clutter the design and create readability issues.
+
+For external Links that aren't in a paragraph or text context, consider [Button's link role](https://gestalt.netlify.app/button#Role) or [IconButton's link role](https://gestalt.netlify.app/iconbutton#Role).`}
```
Comment (dangerismycat) [inline docs/pages/link.js:444]: rewording for clarity: "…consider Button or IconButton with `role="link"`."

File: docs/pages/link.js [hunk 4]
```diff
@@ -624,20 +638,51 @@ However, Link's underline style can be overridden at any time using the \`underl
...
-          <MainSection.Card
-            defaultCode={`
+      </MainSection>
+
+      <MainSection.Subsection
+        title="externalLinkIcon and rel"
+        columns={2}
+        description={`An external link, also called an outbound link, is a link from Pinterest to a different website. External links require specific SEO, visual, and accessibility treatments.
+
+\`rel\` is optional. Use "nofollow" for external links to specify to web crawlers not follow the link. Don't use "nofollow" with urls redirecting to any Pinterest domain or subsite.
```
Comment (dangerismycat) [inline docs/pages/link.js:648]: nit: there's some inconsistency on this page between "subsite" and "sub-site". I don't have an opinion either way, just that we use the same term throughout

File: docs/pages/link.js [hunk 5]
```diff
@@ -624,20 +638,51 @@ However, Link's underline style can be overridden at any time using the \`underl
...
-          description={`\`rel\` is optional. Use "nofollow" for external links to specify to web crawlers not follow the link. Don't use "nofollow" with urls redirecting to any Pinterest site.`}
-        >
-          <MainSection.Card
-            defaultCode={`
+      </MainSection>
+
+      <MainSection.Subsection
+        title="externalLinkIcon and rel"
+        columns={2}
+        description={`An external link, also called an outbound link, is a link from Pinterest to a different website. External links require specific SEO, visual, and accessibility treatments.
```
Comment (dangerismycat) [inline docs/pages/link.js:646]: This section is great. 💯

File: docs/pages/visual-test/Link-dark.js [hunk 6]
```diff
@@ -1,17 +1,32 @@
...
     <ColorSchemeProvider colorScheme="dark">
       <Box color="white" display="inlineBlock" padding={4}>
-        <Text inline>
-          Visit our{' '}
-          <Link inline href="https://pinterest.com">
-            Business Help Center
-          </Link>
-        </Text>
+        <Flex direction="column" gap={2}>
+          ejefcbcvhultvfdhebvtulclubiljucrrkhhljgkditb
```
Comment (dangerismycat) [inline docs/pages/visual-test/Link-dark.js:10]: 😂

File: packages/gestalt/src/Link.js [hunk 7]
```diff
@@ -18,6 +18,59 @@ import getRoundingClassName from './getRoundingClassName.js';
...
+type ExternalLinkIcon =
+  | 'none'
+  | 'default'
+  | {|
+      size: $ElementType<React$ElementConfig<typeof Text>, 'size'>,
+      color: $ElementType<React$ElementConfig<typeof Icon>, 'color'>,
+    |};
+
+function ExternalIcon({
+  externalLinkIcon,
```
Comment (dangerismycat) [inline packages/gestalt/src/Link.js:34]: nit: it's a little odd for a component to take a prop that's essentially the component's name. Within the context of this component, this would be more like just `icon` or something similar

File: packages/gestalt/src/Link.js [hunk 8]
```diff
@@ -18,6 +18,59 @@ import getRoundingClassName from './getRoundingClassName.js';
...
+  position,
+}: {|
+  externalLinkIcon: ExternalLinkIcon,
+  position: 'rtl' | 'ltr',
+|}): Node {
+  const externalLinkIconMap = {
```
Comment (dangerismycat) [inline packages/gestalt/src/Link.js:40]: nit: this can be defined outside of the function so we're not creating a new object every render

File: packages/gestalt/src/Link.js [hunk 9]
```diff
@@ -18,6 +18,59 @@ import getRoundingClassName from './getRoundingClassName.js';
...
+    '300': 16,
+    '400': 20,
+    '500': 28,
+    '600': 36,
+    'sm': 12,
+    'md': 14,
+    'lg': 16,
+  };
+
+  return externalLinkIcon === 'none' ? null : (
```
Comment (dangerismycat) [inline packages/gestalt/src/Link.js:52]: this check isn't necessary, you're already checking at the callsite (or keep this one and remove those)

File: packages/gestalt/src/Link.js [hunk 10]
```diff
@@ -94,6 +147,10 @@ See the [target variant](https://gestalt.pinterest.systems/link#Target) to learn
    * When `underline` is supplied, we override the underline style internally managed by the component. See the [underline variant](https://gestalt.pinterest.systems/link#Underline) to learn more.
    */
   underline?: 'auto' | 'none' | 'always' | 'hover',
+  /**
+   * When supplied, a "visit" icon is shown at the end of Link. See the [externalLinkIcon and rel variant](https://gestalt.pinterest.systems/link#externalLinkIcon-and-rel) to learn more.
+   */
+  externalLinkIcon?: ExternalLinkIcon,
```
Comment (dangerismycat) [inline packages/gestalt/src/Link.js:153]: nit: alphabetize this please

File: packages/gestalt/src/Link.js [hunk 11]
```diff
@@ -18,6 +18,59 @@ import getRoundingClassName from './getRoundingClassName.js';
 import focusStyles from './Focus.css';
 import useFocusVisible from './useFocusVisible.js';
 import layoutStyles from './Layout.css';
+import Icon from './Icon.js';
+import Box from './Box.js';
+import Text from './Text.js';
+
+type ExternalLinkIcon =
```
Comment (dangerismycat) [inline packages/gestalt/src/Link.js:25]: I'm not a huge fan of the polymorphism here, but I also don't have any better ideas. 😬 This works, I think

File: packages/gestalt/src/Link.js [hunk 12]
```diff
@@ -18,6 +18,59 @@ import getRoundingClassName from './getRoundingClassName.js';
...
+    '400': 20,
+    '500': 28,
+    '600': 36,
+    'sm': 12,
+    'md': 14,
+    'lg': 16,
+  };
+
+  return externalLinkIcon === 'none' ? null : (
+    <div aria-hidden className={classnames(position === 'rtl' ? styles.rtlIcon : styles.ltrIcon)}>
```
Comment (dangerismycat) [inline packages/gestalt/src/Link.js:53]: nit: consider some variables for readability:
```js
const isRtl = position === 'rtl';
const isDefault = externalLinkIcon === 'default';

let iconSize = '300';
let iconColor = 'default';
if (!isDefault) {
  // I'm pretty sure Flow will be ok with this, but it might complain
  iconSize = externalLinkIcon?.size ?? iconSize;
  iconColor = externalLinkIcon?.color ?? iconColor;
}

return (
  <div aria-hidden className={classnames(styles[isRtl ? 'rtlIcon' : 'ltrIcon')}>
      <Box
        marginEnd={isRtl ? 1 : 0}
        marginStart={isRtl ? 0 : 1}
        display="inlineBlock"
      >
        <Icon
          inline
          icon="visit"
          accessibilityLabel=""
          size={externalLinkIconMap[iconSize]}
          color={iconColor}
        />
      </Box>
    </div>
  );
```

**Review by dangerismycat — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (AlbertCarreras) [reply CHANGELOG.md:11]: Yes, to prevent
Screen Shot 2022-05-02 at 1 05 51 PM

File: docs/pages/link.js (hunk 3 above)
Comment (AlbertCarreras) [reply docs/pages/link.js:444]: DONE

File: docs/pages/link.js (hunk 2 above)
Comment (AlbertCarreras) [reply docs/pages/link.js:240]: Removed as per conversation with @cintiapub 

Disclaimers are texts like “By continuing, you agree to Pinterest’s Business Terms of Service…”

File: docs/pages/link.js (hunk 4 above)
Comment (AlbertCarreras) [reply docs/pages/link.js:648]: DONE

File: packages/gestalt/src/Link.js (hunk 8 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Link.js:40]: DONE

File: packages/gestalt/src/Link.js (hunk 9 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Link.js:52]: I know you were gonna point at this ones. 

I need them within ExternalIcon for Flow issues. I also put them outside cos I though it'd be more explicit for the parent to control the render or not, but needed the conditional inside as well for Flow.

I'll keep it inside as it's needed.

Comment (AlbertCarreras) [reply packages/gestalt/src/Link.js:52]: DONE

File: packages/gestalt/src/Link.js (hunk 10 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Link.js:153]: DONE

File: packages/gestalt/src/Link.js (hunk 11 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Link.js:25]: I thought this was neater than having 2 props: `isExternal` and `externalIcon`

File: packages/gestalt/src/Link.js (hunk 9 above)
Comment (dangerismycat) [reply packages/gestalt/src/Link.js:46]: Makes sense, Flow needs some convincing sometimes. But consider leaving a comment (in the code) in those situations, so Future Us doesn't waste time trying to refactor it once we've forgotten the context

File: packages/gestalt/src/Link.js (hunk 11 above)
Comment (dangerismycat) [reply packages/gestalt/src/Link.js:37]: Agreed!
